File: src/main/java/com/mparticle/kits/KochavaKit.java
package com.mparticle.kits;

import android.content.Context;
import android.content.Intent;
import android.location.Location;

import com.kochava.base.AttributionUpdateListener;
import com.kochava.base.DeepLinkListener;
import com.kochava.base.Deeplink;
import com.kochava.base.DeeplinkProcessedListener;
import com.kochava.base.ReferralReceiver;
import com.kochava.base.Tracker;
import com.kochava.base.Tracker.Configuration;
import com.mparticle.AttributionError;
import com.mparticle.AttributionResult;
import com.mparticle.MParticle;
import com.mparticle.internal.Logger;
import com.mparticle.internal.MPUtility;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class KochavaKit extends KitIntegration implements KitIntegration.AttributeListener {
    public static final String ATTRIBUTION_PARAMETERS = "attribution";
    public static final String DEEPLINK_PARAMETERS = "deeplink";
    public static final String ENHANCED_DEEPLINK_PARAMETERS = "enhancedDeeplink";

    private static final String APP_ID = "appId";
    private static final String USE_CUSTOMER_ID = "useCustomerId";
    private static final String INCLUDE_ALL_IDS = "passAllOtherIdentities";
    private static final String LIMIT_ADD_TRACKING = "limitAdTracking";
    private static final String RETRIEVE_ATT_DATA = "retrieveAttributionData";
    private static final String ENABLE_LOGGING = "enableLogging";

    private static Map<String, String> identityLink;

    @Override
    public String getName() {
        return "Kochava";
    }

    @Override
    protected List<ReportingMessage> onKitCreate(Map<String, String> settings, Context context) {
        boolean attributionEnabled = Boolean.parseBoolean(getSettings().get(RETRIEVE_ATT_DATA));

        int logLevel = Tracker.LOG_LEVEL_NONE;
        if (Boolean.parseBoolean(getSettings().get(ENABLE_LOGGING))) {
            logLevel = Tracker.LOG_LEVEL_DEBUG;
        }

        Configuration configuration = new Configuration(context.getApplicationContext())
                .setLogLevel(logLevel)
                .setAppGuid(getSettings().get(APP_ID))
                .setAppLimitAdTracking(Boolean.parseBoolean(getSettings().get(LIMIT_ADD_TRACKING)));

        if (identityLink != null) {
            configuration.setIdentityLink(new Tracker.IdentityLink().add(identityLink));
        }
        if (attributionEnabled) {
            configuration.setAttributionUpdateListener(mAttributionListener);
        }

        Tracker.configure(configuration);

        if (attributionEnabled) {
            Tracker.setDeepLinkListener(getKitManager().getLaunchUri(), mDeepLinkListener);
            Tracker.processDeeplink(getKitManager().getLaunchUri().toString(), mDeepLinkProcessedListener);
        }
        return null;
    }

    @Override
    public void setLocation(Location location) {

    }

    @Override
    public void setUserAttribute(String attributeKey, String attributeValue) {

    }

    @Override
    public void setUserAttributeList(String s, List<String> list) {
        
    }

    @Override
    public boolean supportsAttributeLists() {
        return true;
    }

    @Override
    public void setAllUserAttributes(Map<String, String> map, Map<String, List<String>> map1) {

    }

    @Override
    public void removeUserAttribute(String key) {

    }

    @Override
    public void setInstallReferrer(Intent intent) {
        new ReferralReceiver().onReceive(getContext(), intent);
    }

    @Override
    public void setUserIdentity(MParticle.IdentityType identityType, String id) {
        if (identityType == MParticle.IdentityType.CustomerId) {
            if (!getSettings().containsKey(USE_CUSTOMER_ID) ||
                    Boolean.parseBoolean(getSettings().get(USE_CUSTOMER_ID))) {
                Map<String, String> map = new HashMap<>(1);
                map.put(identityType.name(), id);
                Tracker.setIdentityLink(new Tracker.IdentityLink().add(map));
            }
        } else {
            if (Boolean.parseBoolean(getSettings().get(INCLUDE_ALL_IDS))) {
                Map<String, String> map = new HashMap<>(1);
                map.put(identityType.name(), id);
                Tracker.setIdentityLink(new Tracker.IdentityLink().add(map));
            }
        }
    }

    @Override
    public void removeUserIdentity(MParticle.IdentityType identityType) {

    }

    @Override
    public List<ReportingMessage> logout() {
        return null;
    }

    @Override
    public List<ReportingMessage> setOptOut(boolean optOutStatus) {
        Tracker.setAppLimitAdTracking(optOutStatus);
        List<ReportingMessage> messageList = new LinkedList<>();
        messageList.add(
                new ReportingMessage(this, ReportingMessage.MessageType.OPT_OUT, System.currentTimeMillis(), null)
                        .setOptOut(optOutStatus)
        );
        return messageList;
    }

    public static void setIdentityLink(Map<String, String> identityLink) {
        KochavaKit.identityLink = identityLink;
    }

    private void setAttributionResultParameter(String key, JSONObject value) {
        try {
            JSONObject parameters = new JSONObject().put(key, value);
            AttributionResult result = new AttributionResult()
                    .setServiceProviderId(getConfiguration().getKitId())
                    .setParameters(parameters);
            getKitManager().onResult(result);
        } catch (JSONException e) {
            AttributionError error = new AttributionError()
                    .setServiceProviderId(getConfiguration().getKitId())
                    .setMessage(e.getMessage());
            getKitManager().onError(error);
        }
    }

    private AttributionUpdateListener mAttributionListener = new AttributionUpdateListener() {
        @Override
        public void onAttributionUpdated(String s) {
            try {
                JSONObject attributionJson = new JSONObject(s);
                setAttributionResultParameter(ATTRIBUTION_PARAMETERS, attributionJson);
            } catch (JSONException e) {
                AttributionError error = new AttributionError()
                        .setMessage("unable to parse attribution JSON:\n " + s)
                        .setServiceProviderId(getConfiguration().getKitId());
                getKitManager().onError(error);
            }
        }
    };

    private DeepLinkListener mDeepLinkListener = new DeepLinkListener() {
        @Override
        public void onDeepLink(Map<String, String> map) {
            if (!MPUtility.isEmpty(map)) {
                setAttributionResultParameter(DEEPLINK_PARAMETERS, MPUtility.mapToJson(map));
            }
        }
    };

    private DeeplinkProcessedListener mDeepLinkProcessedListener = new DeeplinkProcessedListener() {
        @Override
        public void onDeeplinkProcessed(Deeplink deeplink) {
            if (deeplink != null) {
                setAttributionResultParameter(ENHANCED_DEEPLINK_PARAMETERS, deeplink.toJson());
            }
        }
    };
}

File: src/main/kotlin/com/mparticle/kits/KochavaKit.kt
package com.mparticle.kits

import android.content.Context
import com.mparticle.kits.KitIntegration.AttributeListener
import com.kochava.base.Tracker
import com.kochava.base.Tracker.IdentityLink
import android.content.Intent
import android.location.Location
import com.kochava.base.ReferralReceiver
import com.mparticle.MParticle.IdentityType
import java.util.HashMap
import org.json.JSONObject
import com.mparticle.AttributionResult
import org.json.JSONException
import com.mparticle.AttributionError
import com.kochava.base.AttributionUpdateListener
import com.kochava.base.DeepLinkListener
import com.mparticle.internal.MPUtility
import com.kochava.base.DeeplinkProcessedListener

class KochavaKit : KitIntegration(), AttributeListener {
    override fun getName(): String = NAME

    override fun onKitCreate(
        settings: Map<String, String>,
        context: Context
    ): List<ReportingMessage>? {
        val attributionEnabled = java.lang.Boolean.parseBoolean(getSettings()[RETRIEVE_ATT_DATA])
        var logLevel = Tracker.LOG_LEVEL_NONE
        if (java.lang.Boolean.parseBoolean(getSettings()[ENABLE_LOGGING])) {
            logLevel = Tracker.LOG_LEVEL_DEBUG
        }
        val configuration = getSettings()[APP_ID]?.let {
            Tracker.Configuration(context.applicationContext)
                .setLogLevel(logLevel)
                .setAppGuid(it)
                .setAppLimitAdTracking(java.lang.Boolean.parseBoolean(getSettings()[LIMIT_ADD_TRACKING]))
        }
        if (configuration != null) {
            identityLink?.let {
                configuration.setIdentityLink(IdentityLink().add(it))
            }
            if (attributionEnabled) {
                configuration.setAttributionUpdateListener(mAttributionListener)
            }
            Tracker.configure(configuration)
            if (attributionEnabled) {
                Tracker.setDeepLinkListener(kitManager.launchUri, mDeepLinkListener)
                Tracker.processDeeplink(kitManager.launchUri.toString(), mDeepLinkProcessedListener)
            }
        }
        return null
    }

    override fun setLocation(location: Location) {}
    override fun setUserAttribute(attributeKey: String, attributeValue: String) {}
    override fun setUserAttributeList(s: String, list: List<String>) {}
    override fun supportsAttributeLists(): Boolean = true
    override fun setAllUserAttributes(map: Map<String, String>, map1: Map<String, List<String>>) {}
    override fun removeUserAttribute(key: String) {}
    override fun setInstallReferrer(intent: Intent) {
        ReferralReceiver().onReceive(context, intent)
    }

    override fun setUserIdentity(identityType: IdentityType, id: String) {
        if (identityType == IdentityType.CustomerId) {
            if (!settings.containsKey(USE_CUSTOMER_ID) ||
                (settings[USE_CUSTOMER_ID].toBoolean())
            ) {
                val map = HashMap<String, String>(1)
                map[identityType.name] = id
                Tracker.setIdentityLink(IdentityLink().add(map))
            }
        } else {
            if ((settings[INCLUDE_ALL_IDS]).toBoolean()) {
                val map = HashMap<String, String>(1)
                map[identityType.name] = id
                Tracker.setIdentityLink(IdentityLink().add(map))
            }
        }
    }

    override fun removeUserIdentity(identityType: IdentityType) {}
    override fun logout(): List<ReportingMessage> = emptyList()

    override fun setOptOut(optOutStatus: Boolean): List<ReportingMessage> {
        Tracker.setAppLimitAdTracking(optOutStatus)

        return listOf(
            ReportingMessage(
                this,
                ReportingMessage.MessageType.OPT_OUT,
                System.currentTimeMillis(),
                null
            ).setOptOut(optOutStatus)
        )
    }

    private fun setAttributionResultParameter(key: String, value: JSONObject) {
        try {
            val parameters = JSONObject().put(key, value)
            val result = AttributionResult()
                .setServiceProviderId(configuration.kitId)
                .setParameters(parameters)
            kitManager.onResult(result)
        } catch (e: JSONException) {
            val error = AttributionError()
                .setServiceProviderId(configuration.kitId)
                .setMessage(e.message)
            kitManager.onError(error)
        }
    }

    private val mAttributionListener = AttributionUpdateListener { s ->
        try {
            val attributionJson = JSONObject(s)
            setAttributionResultParameter(ATTRIBUTION_PARAMETERS, attributionJson)
        } catch (e: JSONException) {
            val error = AttributionError()
                .setMessage("unable to parse attribution JSON:\n $s")
                .setServiceProviderId(configuration.kitId)
            kitManager.onError(error)
        }
    }
    private val mDeepLinkListener = DeepLinkListener { map ->
        if (!MPUtility.isEmpty(map)) {
            setAttributionResultParameter(DEEPLINK_PARAMETERS, MPUtility.mapToJson(map))
        }
    }
    private val mDeepLinkProcessedListener = DeeplinkProcessedListener { deeplink ->
        setAttributionResultParameter(ENHANCED_DEEPLINK_PARAMETERS, deeplink.toJson())
    }

    companion object {
        const val ATTRIBUTION_PARAMETERS = "attribution"
        const val DEEPLINK_PARAMETERS = "deeplink"
        const val ENHANCED_DEEPLINK_PARAMETERS = "enhancedDeeplink"
        private const val APP_ID = "appId"
        private const val USE_CUSTOMER_ID = "useCustomerId"
        private const val INCLUDE_ALL_IDS = "passAllOtherIdentities"
        private const val LIMIT_ADD_TRACKING = "limitAdTracking"
        private const val RETRIEVE_ATT_DATA = "retrieveAttributionData"
        private const val ENABLE_LOGGING = "enableLogging"
        const val NAME = "Kochava"
        private var identityLink: Map<String, String>? = null
        fun setIdentityLink(identityLink: Map<String, String>?) {
            Companion.identityLink = identityLink
        }
    }
}
